### PR TITLE
Use `torch.no_grad()`, `model.eval()` and freeze models

### PR DESCRIPTION
**Pull request by jopperm**
Tailor benchmarking more towards inference, by consistently using `torch.no_grad()`, calling `model.eval()` and allowing inductor to treat weights as constants (a.k.a. freezing). For the pytorch-compile backend on CPU, this enables oneDNN post-op fusion and the use of the `TORCHINDUCTOR_MAX_AUTOTUNE=1` mode in which a custom fused GEMM is generated from a C++/OpenMP template.

### CODE REVIEW

File: ai_bench/harness/runner/kernel_runner.py [hunk 1]
```diff
@@ -243,14 +244,15 @@ def benchmark_model(self, variant, model, args) -> KernelStats:
...
         # Measure performance.
-        meas_us = testing.time(
-            fn,
-            args,
-            warmup=self.warmup,
-            rep=self.rep,
-            min_cache_nuke_mib=self.min_cache_nuke_mib,
-            device=self.device,
-        )
+        with torch.no_grad(), inductor_config.patch({"freezing": True}):
```
Comment (adam-smnk) [inline ai_bench/harness/runner/kernel_runner.py]: I wonder if it's not too late to set freezing just before running as I'm not sure if there are any configs propagated earlier.
Maybe it should be set already when `model.compile` is called or as a global setting instead of temporary patch?

**Review by adam-smnk — APPROVED** [review]
Good idea 👍
Wonder if we should call `model.eval()` too but never dug deeper into that.

**Comment by jopperm** [comment]
> Wonder if we should call `model.eval()` too but never dug deeper into that.

Good point, yeah I think we should. I saw it recommended as well but didn't include it initially because it didn't matter for the simple fused matmuls I was looking at today. There are a couple of KB problems using dropout, and these won't validate in train mode because of their inherent randomness.

File: ai_bench/harness/runner/kernel_runner.py (hunk 1 above)
Comment (jopperm) [reply ai_bench/harness/runner/kernel_runner.py]: Changed it; setting it globally is probably more robust. One can still control freezing via inductor's own env (`TORCHINDUCTOR_FREEZING`).
